runtime-sdk/src/modules/rofl/app/client.rs

```diff
@@ -49,28 +69,148 @@ where
+        self.imp.estimate_gas(req).await
+    }
+
+    /// Sign a given transaction, submit it and wait for block inclusion.
+    ///
+    /// This method supports multiple transaction signers.
+    pub async fn multi_sign_and_submit_tx(
+        &self,
+        signers: &[Arc<dyn Signer>],
+        tx: transaction::Transaction,
+    ) -> Result<transaction::CallResult> {
+        self.multi_sign_and_submit_tx_opts(SubmitTxOpts::default(), signers, tx)
+            .await
+    }
+
+    /// Sign a given transaction, submit it and wait for block inclusion.
+    ///
+    /// This method supports multiple transaction signers.
+    pub async fn multi_sign_and_submit_tx_opts(
+        &self,
+        opts: SubmitTxOpts,
+        signers: &[Arc<dyn Signer>],
+        tx: transaction::Transaction,
+    ) -> Result<transaction::CallResult> {
+        self.submission_mgr
+            .multi_sign_and_submit_tx(opts, signers, tx)
+            .await
+    }
+
+    /// Sign a given transaction, submit it and wait for block inclusion.
+    pub async fn sign_and_submit_tx(
+        &self,
+        signer: Arc<dyn Signer>,
+        tx: transaction::Transaction,
+    ) -> Result<transaction::CallResult> {
+        self.multi_sign_and_submit_tx(&[signer], tx).await
+    }
+
+    /// Run a closure inside a `CurrentState` context with store for the given round.
+    pub async fn with_store_for_round<F, R>(&self, round: u64, f: F) -> Result<R>
+    where
+        F: FnOnce() -> Result<R> + Send + 'static,
+        R: Send + 'static,
+    {
+        self.imp.with_store_for_round(round, f).await
+    }
+
+    /// Return a store corresponding to the given round.
+    pub async fn store_for_round(&self, round: u64) -> Result<HostStore> {
+        self.imp.store_for_round(round).await
+    }
+}
+
+impl<A> Clone for Client<A>
+where
+    A: App,
+{
+    fn clone(&self) -> Self {
+        Self {
+            imp: self.imp.clone(),
+            submission_mgr: self.submission_mgr.clone(),
+        }
+    }
+}
+
+struct ClientImpl<A: App> {
+    state: Arc<processor::State<A>>,
+    cmdq: mpsc::WeakSender<processor::Command>,
+    latest_round: Arc<AtomicU64>,
+}
+
+impl<A> ClientImpl<A>
+where
+    A: App,
+{
+    fn new(state: Arc<processor::State<A>>, cmdq: mpsc::WeakSender<processor::Command>) -> Self {
+        Self {
+            state,
+            cmdq,
+            latest_round: Arc::new(AtomicU64::new(0)),
+        }
+    }
+
+    /// Retrieve the latest known runtime round.
+    async fn latest_round(&self) -> Result<u64> {
         let cmdq = self
             .cmdq
             .upgrade()
             .ok_or(anyhow!("processor has shut down"))?;
         let (tx, rx) = oneshot::channel();
         cmdq.send(processor::Command::GetLatestRound(tx)).await?;
-        Ok(rx.await?)
+        let round = rx.await?;
+        Ok(self
+            .latest_round
+            .fetch_max(round, Ordering::SeqCst)
+            .max(round))
```

Comment:
Why is this needed? We always fetch the latest round, so it is highly unlikely that this `max` will ever return `self.latest_round`.

Reply:
To avoid possible race conditions when the runtime hasn't yet been notified of the new round but transaction submission has just completed and we know that the round has advanced.